File: SeaCleaner/Client/Game/Trash.cs
using Microsoft.JSInterop;
using System;
using System.Threading.Tasks;

namespace SeaCleaner.Client.Game
{
    internal class BoundingBox
    {
        public double PosX { get; set; }
        public double PosY { get; set; }
        public int Width { get; private set; }
        public int Height { get; private set; }

        public BoundingBox(double x, double y, int width, int height)
        {
            PosX = x;
            PosY = y;
            Width = width;
            Height = height;
        }

        public bool IsPointInside(double pointX, double pointY) =>
           (pointX >= PosX) && (pointX <= PosX + Width) && (pointY >= PosY) && (pointY <= PosY + Height);

        public bool IsIntersect(BoundingBox other) =>
            IsPointInside(other.PosX, other.PosY) ||
            IsPointInside(other.PosX + other.Width, other.PosY) ||
            IsPointInside(other.PosX, other.PosY + other.Height) ||
            IsPointInside(other.PosX + other.Width, other.PosY + other.Height);
    }

    internal class Trash
    {
        private readonly SpriteImageInfo _imageInfo;
        public BoundingBox BBox { get; private set; }
        public Trash(SpriteImageInfo imageInfo)
        {
            _imageInfo = imageInfo;
            _random = new Random();
            _posX = Math.Round(_random.NextDouble() * 1000) % 800 + 100;
            _posYWeggle = Math.Round(_random.NextDouble() * 1000) % 450 + 200;
            _posYStatic = _posYWeggle;
            _fCounter = Math.Round(_random.NextDouble() * 10000) % 1800;
            BBox = new BoundingBox(_posX, _posYWeggle, _imageInfo.SpriteWidth, _imageInfo.SpriteHeight);
        }

        private readonly Random _random;
        private double _posX;
        private double _posYWeggle;
        private double _posYStatic;
        private readonly double _fluctCoef = 0.1;
        private readonly double _fluctuation = 0.5;
        private double _fCounter;
        private int _counter = 0;

        public bool IsCaptured { get; set; }
        public bool IsVisible { get; set; } = true;
    
        public double PosX
        {
            get => _posX;
            set
            {
                _posX = value;
                BBox.PosX = value;
            }
        }

        public double PosY
        {
            get => _posYWeggle;
            set
            {
                _posYWeggle = value;
                _posYStatic = value;
                BBox.PosY = value;
            }
        } 

        public void Update()
        {
            if (IsCaptured) return;

            if (++_counter >= 2)
            {
                _fCounter++;
                if (_fCounter >= 1800)
                {
                    _fCounter = 0;
                    _posYWeggle = _posYStatic; // для корректировки положения, которое отклоняется из-за округлений при расчетах покачивания
                }
                _counter = 0;
            }

            if (!IsCaptured)
            {
                // покачивание вверх-вниз
                _posYWeggle += Math.Sin(_fCounter * _fluctCoef) * _fluctuation;
                BBox.PosY = _posYWeggle;
            }
        }

        public async ValueTask Draw(IJSRuntime jsRuntime)
        {
            if (IsVisible)
            {
                await jsRuntime.InvokeVoidAsync("drawImage", _imageInfo.SpriteName, PosX, PosY, _imageInfo.SpriteWidth, _imageInfo.SpriteHeight);
                if (Game.SHOW_BB)
                {
                    await jsRuntime.InvokeVoidAsync("drawRect", BBox.PosX, BBox.PosY, BBox.Width, BBox.Height);
                }
            }
        }
    }
}

File: SeaCleaner/Client/wwwroot/js/Trash.js
/* 
 * класс для обработки столкновений
 */
function CBoundingBox(_X, _Y, _W, _H){
    this.PosX = _X;
    this.PosY = _Y;
    this.width = _W;
    this.height = _H;
    this.isPointInside = function(_X, _Y){
        if(((_X >= this.PosX) && (_X <= this.PosX + this.width)) &&
           ((_Y >= this.PosY) && (_Y <= this.PosY + this.height)))
            return true;
        else
            return false;
    };
    this.isIntersect = function(Bbox){
        // пересечение двух прямоугольных областей
        if(this.isPointInside(Bbox.PosX, Bbox.PosY) ||
           this.isPointInside(Bbox.PosX + Bbox.width, Bbox.PosY) ||
           this.isPointInside(Bbox.PosX, Bbox.PosY + Bbox.height)||
           this.isPointInside(Bbox.PosX + Bbox.width, Bbox.PosY + Bbox.height))
            return true;
        else
            return false;   
    };
}
/*
 * класс для мусора
 */
function CTrash(_SpriteName){
    this.spriteName = _SpriteName;
    this.PosX = Math.round(Math.random()*1000) % 800 + 100;
    this.PosY = Math.round(Math.random()*1000) % 450 + 200;
    this.PosY_ = this.PosY;
    this.fluctCoef = 0.1;
    this.fluctuation = 0.5;
    this.isCaptured = false;
    this.isVisible = true;
    this.counter = 0;
    this.__defineSetter__("posX", function(value){
        this.PosX = value;
        this.bBox.PosX = value;
    });
    this.__defineSetter__("posY", function(value){
        this.PosY = value;
        this.bBox.PosY = value;
        this.PosY_ = value;
    });
    this.initialize = function(){
        this.fCounter = Math.round(Math.random() * 10000) % 1800;
        this.width = MultiImage[this.spriteName].width;
        this.height = MultiImage[this.spriteName].height;
        this.bBox = new CBoundingBox(this.PosX, this.PosY, this.width, this.height);
    };
    this.update = function(){
        if(this.isCaptured) return;
        //
        if(++this.counter >= 2){
        
            this.fCounter++;
            if(this.fCounter === 1800){
                this.fCounter = 0;
                this.PosY = this.PosY_;
            }
            this.counter = 0;
        }
        if(!this.isCaptured){
            this.PosY += Math.sin(this.fCounter * this.fluctCoef) * this.fluctuation;
            this.bBox.PosY = this.PosY;
        }
    };
    this.draw = function(context){
        if(this.isVisible){
            context.drawImage( 
                MultiImage[this.spriteName], 
                this.PosX, this.PosY,  
                this.width, this.height);              
        }    
    };
}

